File: backend/app/services/status_map.py
"""Map carrier-specific status vocab onto NormalizedStatus. Tables are data so they're easy to extend."""

from __future__ import annotations

import re

from app.enums import NormalizedStatus as S

# ---------------------------------------------------------------- USPS
USPS_CATEGORY: dict[str, S] = {
    "pre-shipment": S.LABEL_CREATED,
    "preshipment": S.LABEL_CREATED,
    "shipping label created": S.LABEL_CREATED,
    "in transit": S.IN_TRANSIT,
    "accepted": S.IN_TRANSIT,
    "moving through network": S.IN_TRANSIT,
    "arriving early": S.IN_TRANSIT,
    "arriving late": S.IN_TRANSIT,
    "arriving on time": S.IN_TRANSIT,
    "out for delivery": S.OUT_FOR_DELIVERY,
    "delivered": S.DELIVERED,
    "delivered to agent": S.DELIVERED,
    "available for pickup": S.IN_TRANSIT,
    "alert": S.EXCEPTION,
    "return to sender": S.RETURNED,
    "returned": S.RETURNED,
}

USPS_ATTENTION_CATEGORY = {"available for pickup": "pickup"}

# USPS Publication 97 / TextTrack event codes (subset)
USPS_EVENT_CODE: dict[str, S] = {
    "GX": S.LABEL_CREATED,
    "MA": S.LABEL_CREATED,
    "GS": S.LABEL_CREATED,
    "03": S.IN_TRANSIT,
    "07": S.IN_TRANSIT,
    "10": S.IN_TRANSIT,
    "T1": S.IN_TRANSIT,
    "TM": S.IN_TRANSIT,
    "SF": S.IN_TRANSIT,
    "PC": S.IN_TRANSIT,
    "AE": S.IN_TRANSIT,
    "AD": S.IN_TRANSIT,
    "A1": S.IN_TRANSIT,
    "NT": S.IN_TRANSIT,
    "EF": S.IN_TRANSIT,
    "OA": S.IN_TRANSIT,
    "L1": S.IN_TRANSIT,
    "OF": S.OUT_FOR_DELIVERY,
    "59": S.OUT_FOR_DELIVERY,
    "01": S.DELIVERED,
    "DX": S.DELIVERED,
    "DN": S.DELIVERED,
    "17": S.DELIVERED,
    "16": S.IN_TRANSIT,  # available for pickup
    "02": S.EXCEPTION,
    "04": S.EXCEPTION,
    "05": S.EXCEPTION,
    "06": S.EXCEPTION,
    "53": S.EXCEPTION,
    "55": S.EXCEPTION,
    "56": S.EXCEPTION,
    "H0": S.EXCEPTION,
    "51": S.EXCEPTION,
    "52": S.EXCEPTION,
    "09": S.RETURNED,
    "21": S.RETURNED,
    "28": S.RETURNED,
    "29": S.RETURNED,
    "31": S.RETURNED,
}
USPS_ATTENTION_CODE = {
    "16": "pickup",
    "02": "delivery_failed",
    "55": "delivery_failed",
    "53": "delivery_failed",
}

# ---------------------------------------------------------------- FedEx
FEDEX_CODE: dict[str, S] = {
    "OC": S.LABEL_CREATED,
    "IN": S.LABEL_CREATED,
    "PU": S.IN_TRANSIT,
    "IT": S.IN_TRANSIT,
    "AR": S.IN_TRANSIT,
    "DP": S.IN_TRANSIT,
    "AF": S.IN_TRANSIT,
    "CD": S.IN_TRANSIT,
    "CC": S.IN_TRANSIT,
    "HL": S.IN_TRANSIT,
    "PF": S.IN_TRANSIT,
    "PM": S.IN_TRANSIT,
    "SP": S.IN_TRANSIT,
    "PX": S.IN_TRANSIT,
    "SE": S.EXCEPTION,
    "FD": S.IN_TRANSIT,
    "TR": S.IN_TRANSIT,
    "OD": S.OUT_FOR_DELIVERY,
    "DL": S.DELIVERED,
    "DE": S.EXCEPTION,
    "DY": S.EXCEPTION,
    "CA": S.EXCEPTION,
    "RR": S.EXCEPTION,
    "RS": S.RETURNED,
    "RG": S.RETURNED,
    "RP": S.RETURNED,
}
FEDEX_ATTENTION_CODE = {"HL": "pickup", "DE": "delivery_failed", "RR": "delivery_failed"}

# ---------------------------------------------------------------- keyword fallback (any carrier)
KEYWORDS: list[tuple[re.Pattern[str], S]] = [
    (re.compile(r"\bdelivered\b(?!.*\bnot\b)", re.I), S.DELIVERED),
    (re.compile(r"out for delivery", re.I), S.OUT_FOR_DELIVERY),
    (re.compile(r"return(ed|ing)? to (sender|shipper)|returned", re.I), S.RETURNED),
    (
        re.compile(
            r"label created|shipping label|awaiting item|pre-?shipment|shipment information sent", re.I
        ),
        S.LABEL_CREATED,
    ),
    (
        re.compile(
            r"notice left|undeliverable|refused|delay|alert|held|exception|damaged|unable|missed|attempt",
            re.I,
        ),
        S.EXCEPTION,
    ),
    (re.compile(r"available for pickup|pickup", re.I), S.IN_TRANSIT),
    (
        re.compile(r"in transit|arrived|departed|accepted|processed|picked up|on its way|moving", re.I),
        S.IN_TRANSIT,
    ),
]


def by_keywords(text: str | None) -> S | None:
    if not text:
        return None
    for pat, status in KEYWORDS:
        if pat.search(text):
            return status
    return None


def map_usps(category: str | None, latest_code: str | None, description: str | None) -> tuple[S, str | None]:
    """Returns (status, attention_flag)."""
    cat = (category or "").strip().lower()
    if cat in USPS_CATEGORY:
        return USPS_CATEGORY[cat], USPS_ATTENTION_CATEGORY.get(cat)
    code = (latest_code or "").strip().upper()
    if code in USPS_EVENT_CODE:
        return USPS_EVENT_CODE[code], USPS_ATTENTION_CODE.get(code)
    for key, status in USPS_CATEGORY.items():
        if key in cat:
            return status, USPS_ATTENTION_CATEGORY.get(key)
    kw = by_keywords(description) or by_keywords(category)
    if kw:
        return kw, None
    return S.UNKNOWN, None


def map_fedex(derived_code: str | None, code: str | None, description: str | None) -> tuple[S, str | None]:
    for c in (derived_code, code):
        c = (c or "").strip().upper()
        if c in FEDEX_CODE:
            return FEDEX_CODE[c], FEDEX_ATTENTION_CODE.get(c)
    kw = by_keywords(description)
    if kw:
        return kw, None
    return S.UNKNOWN, None

File: frontend/src/local/statusMap.ts
/** Carrier status vocabularies -> normalized status. Port of backend/app/services/status_map.py. */
import type { Status } from '@/local/carriers'

export const USPS_CATEGORY: Record<string, Status> = {
  'pre-shipment': 'label_created', preshipment: 'label_created', 'shipping label created': 'label_created',
  'in transit': 'in_transit', accepted: 'in_transit', 'moving through network': 'in_transit', 'arriving early': 'in_transit',
  'arriving late': 'in_transit', 'arriving on time': 'in_transit', 'out for delivery': 'out_for_delivery', delivered: 'delivered',
  'delivered to agent': 'delivered', 'available for pickup': 'in_transit', alert: 'exception', 'return to sender': 'returned', returned: 'returned',
}
const USPS_ATTENTION_CATEGORY: Record<string, string> = { 'available for pickup': 'pickup' }

export const USPS_EVENT_CODE: Record<string, Status> = {
  GX: 'label_created', MA: 'label_created', GS: 'label_created',
  '03': 'in_transit', '07': 'in_transit', '10': 'in_transit', T1: 'in_transit', TM: 'in_transit', SF: 'in_transit', PC: 'in_transit',
  AE: 'in_transit', AD: 'in_transit', A1: 'in_transit', NT: 'in_transit', EF: 'in_transit', OA: 'in_transit', L1: 'in_transit',
  OF: 'out_for_delivery', '59': 'out_for_delivery',
  '01': 'delivered', DX: 'delivered', DN: 'delivered', '17': 'delivered',
  '16': 'in_transit',
  '02': 'exception', '04': 'exception', '05': 'exception', '06': 'exception', '53': 'exception', '55': 'exception', '56': 'exception',
  H0: 'exception', '51': 'exception', '52': 'exception',
  '09': 'returned', '21': 'returned', '28': 'returned', '29': 'returned', '31': 'returned',
}
const USPS_ATTENTION_CODE: Record<string, string> = { '16': 'pickup', '02': 'delivery_failed', '55': 'delivery_failed', '53': 'delivery_failed' }

export const FEDEX_CODE: Record<string, Status> = {
  OC: 'label_created', IN: 'label_created',
  PU: 'in_transit', IT: 'in_transit', AR: 'in_transit', DP: 'in_transit', AF: 'in_transit', CD: 'in_transit', CC: 'in_transit', HL: 'in_transit',
  PF: 'in_transit', PM: 'in_transit', SP: 'in_transit', PX: 'in_transit', SE: 'exception', FD: 'in_transit', TR: 'in_transit',
  OD: 'out_for_delivery', DL: 'delivered',
  DE: 'exception', DY: 'exception', CA: 'exception', RR: 'exception',
  RS: 'returned', RG: 'returned', RP: 'returned',
}
const FEDEX_ATTENTION_CODE: Record<string, string> = { HL: 'pickup', DE: 'delivery_failed', RR: 'delivery_failed' }

const KEYWORDS: [RegExp, Status][] = [
  [/\bdelivered\b(?!.*\bnot\b)/i, 'delivered'],
  [/out for delivery/i, 'out_for_delivery'],
  [/return(ed|ing)? to (sender|shipper)|returned/i, 'returned'],
  [/label created|shipping label|awaiting item|pre-?shipment|shipment information sent/i, 'label_created'],
  [/notice left|undeliverable|refused|delay|alert|held|exception|damaged|unable|missed|attempt/i, 'exception'],
  [/available for pickup|pickup/i, 'in_transit'],
  [/in transit|arrived|departed|accepted|processed|picked up|on its way|moving/i, 'in_transit'],
]

export function byKeywords(text: string | null | undefined): Status | null {
  if (!text) return null
  for (const [re, s] of KEYWORDS) if (re.test(text)) return s
  return null
}

/** Returns [status, attention_flag]. */
export function mapUsps(category: string | null | undefined, latestCode: string | null | undefined, description: string | null | undefined): [Status, string | null] {
  const cat = (category ?? '').trim().toLowerCase()
  if (cat in USPS_CATEGORY) return [USPS_CATEGORY[cat], USPS_ATTENTION_CATEGORY[cat] ?? null]
  const code = (latestCode ?? '').trim().toUpperCase()
  if (code in USPS_EVENT_CODE) return [USPS_EVENT_CODE[code], USPS_ATTENTION_CODE[code] ?? null]
  for (const [key, status] of Object.entries(USPS_CATEGORY)) if (cat.includes(key)) return [status, USPS_ATTENTION_CATEGORY[key] ?? null]
  const kw = byKeywords(description) ?? byKeywords(category)
  if (kw) return [kw, null]
  return ['unknown', null]
}

export function mapFedex(derivedCode: string | null | undefined, code: string | null | undefined, description: string | null | undefined): [Status, string | null] {
  for (const c of [derivedCode, code]) {
    const u = (c ?? '').trim().toUpperCase()
    if (u in FEDEX_CODE) return [FEDEX_CODE[u], FEDEX_ATTENTION_CODE[u] ?? null]
  }
  const kw = byKeywords(description)
  if (kw) return [kw, null]
  return ['unknown', null]
}
